# 1.1.2 线程的终止

## stop
错误的线程终止方法stop会产生线程安全问题，stop会强行停止线程的运行，从而产生不可预期的问题。
```java
public class Demo3 {
    public static void main(String[] args) throws InterruptedException {
        StopThread thread = new StopThread();
        thread.start();
        // 休眠1秒，确保i变量自增成功
        Thread.sleep(1000);
        // 暂停线程
          thread.stop(); // 错误的终止
        thread.interrupt(); // 正确终止
        while (thread.isAlive()) {
            // 确保线程已经终止
        } // 输出结果
        thread.print();
    }
}

public class StopThread extends Thread {
  private int i = 0, j = 0;

  @Override
  public void run() {
    synchronized (this) {
	    // 增加同步锁，确保线程安全
	    ++i;
	    try {
	      // 休眠10秒,模拟耗时操作
	      Thread.sleep(10000);
	    } catch (InterruptedException e) {
	      e.printStackTrace();
	    }
	    ++j;
    }
  }
  // 输出结果是
  // i=1 j=0 只有i正确执行了，j在执行之前被强行终止，并且没有任何警告信息。
```
## interrupt
如果目标线程在调用Object class的wait() wait(long) wait(long,int) join() join(long,int) sleep(long,int) 方法被阻塞，那么interrupt会生效，该线程的终端状态将被清除，抛出InterruptException异常。

如果目标线程是被I/O或者是NIO中的Channel所阻塞，I/O操作会被终端或者放回特殊的异常值。达到终止线程的目的。

如果都不满足，则会设置此线程为中断状态。

上面的Demo3以interrupt方式来执行。
输出的结果就是 i=1 j=1 同时会得到一个异常提醒。运算的结果是符合预期的。

## 标志位
如果两个线程同时共享一个标志位来做循环的条件也拿来做为线程终止的信号。
```java
/** 通过状态位来判断 */
public class Demo4 extends Thread {
  public volatile static boolean flag = true;

  public static void main(String[] args) throws InterruptedException {
    new Thread(() -> {
      try {
        while (flag) { // 判断是否运行
          System.out.println("运行中");
          Thread.sleep(1000L);
        }
      } catch (InterruptedException e) {
        e.printStackTrace();
      }
    }).start();
    // 3秒之后，将状态标志改为False，代表不继续运行
    Thread.sleep(3000L);
    flag = false;
    System.out.println("程序运行结束");
  }
}
// 输出结果为
运行中
运行中
运行中
程序运行结束
```



